trieste/acquisition/rule.py

```diff
@@ -1103,6 +1115,380 @@ def state_func(
+            rule = EfficientGlobalOptimization()
+
+        self._subspace_type = subspace_type
+        self._rule = rule
+        self._number_of_tr = number_of_tr
+        self._tags = tuple([str(index) for index in range(number_of_tr)])
+
+    def __repr__(self) -> str:
+        """"""
+        return f"""MultiTrustRegion(
+            {self._subspace_type!r},
+            {self._rule!r},
+            {self._number_of_tr!r})"""
+
+    def acquire(
+        self,
+        search_space: SearchSpace,
+        models: Mapping[Tag, ProbabilisticModelType],
+        datasets: Optional[Mapping[Tag, Dataset]] = None,
+    ) -> types.State[State | None, TensorType]:
+        def state_func(
+            state: MultiTrustRegion.State | None,
+        ) -> Tuple[MultiTrustRegion.State | None, TensorType]:
+            """If state is None, initialise the subspaces by picking new locations. Otherwise,
+            update the existing subspaces.
+            Reinitialise the subspaces if necessary, potentially looking at the entire group.
+            Use the rule to acquire points from the acquisition space.
+            """
+            # If state is set, the tags should be the same as the tags of the acquisition space
+            # in the state.
+            if state is not None:
+                assert (
+                    self._tags == state.acquisition_space.subspace_tags
+                ), f"""The tags of the state acquisition space
+                    {state.acquisition_space.subspace_tags} should be the same as the tags of the
+                    MultiTrustRegion acquisition rule {self._tags}"""
+
+            subspaces = []
+            for tag in self._tags:
+                if state is None:
+                    subspace = self.create_subspace(search_space)
+                    subspace.reinitialise(models, datasets)
+                else:
+                    _subspace = state.acquisition_space.get_subspace(tag)
+                    assert isinstance(_subspace, self._subspace_type)
+                    subspace = _subspace
+                    subspace.update(models, datasets)
+
+                subspaces.append(subspace)
+
+            self.maybe_reinitialise_subspaces(subspaces, models, datasets)
+
+            if state is None:
+                acquisition_space = TaggedMultiSearchSpace(subspaces, self._tags)
+            else:
+                acquisition_space = state.acquisition_space
+
+            state_ = MultiTrustRegion.State(acquisition_space)
+            points = self._rule.acquire(acquisition_space, models, datasets=datasets)
+
+            return state_, points
+
+        return state_func
+
+    def create_subspace(
+        self,
+        search_space: SearchSpace,
+    ) -> UpdateableSearchSpaceType:
+        """Create a subspace from the given search space.
+        This is the default implementation. Can be overridden by subclasses."""
+        return self._subspace_type()
+
+    def maybe_reinitialise_subspaces(
```

Comment:
so If I have something to compute that would be used in reinitializing each TR (e.g. location candidates), I would have to change this method?
I think this would be a common pattern, but then there would be some extra objects to pass to `reinitialize` method below

Reply:
Yes that is correct. I'll add kwargs to both `UpdateableSearchSpace.reinitialize` and `UpdateableSearchSpace.update`.